crates/runtime/src/lib.rs

```diff
@@ -191,23 +194,76 @@ pub struct Runtime {
     pub pods_watcher: Option<podswatcher::PodsWatcher>,
     pub secrets_provider: Arc<RwLock<secrets::SecretsProvider>>,
 
+    extensions: Arc<Vec<RwLock<Box<dyn Extension>>>>,
     spaced_tracer: Arc<tracers::SpacedTracer>,
 }
 
 impl Runtime {
     #[must_use]
-    pub async fn new(app: Option<app::App>, df: Arc<RwLock<DataFusion>>) -> Self {
+    pub async fn new(
+        app: Option<app::App>,
+        df: Arc<RwLock<DataFusion>>,
+        extension_factories: Arc<Vec<Box<dyn ExtensionFactory>>>,
+    ) -> Self {
         dataconnector::register_all().await;
         dataaccelerator::register_all().await;
 
-        Runtime {
+        let mut rt = Runtime {
             app: Arc::new(RwLock::new(app)),
             df,
             models: Arc::new(RwLock::new(HashMap::new())),
             llms: Arc::new(RwLock::new(HashMap::new())),
             pods_watcher: None,
             secrets_provider: Arc::new(RwLock::new(secrets::SecretsProvider::new())),
             spaced_tracer: Arc::new(tracers::SpacedTracer::new(Duration::from_secs(15))),
+            extensions: Arc::new(vec![]),
+        };
+
+        let mut extensions: Vec<RwLock<Box<dyn Extension>>> = vec![];
+        for factory in extension_factories.iter() {
+            let extension = factory.create();
+            let extension_name = extension.name();
+            let extension_lock = RwLock::new(extension);
+            if let Err(err) = rt.initialize_extension(&extension_lock).await {
+                tracing::warn!("Failed to initialize extension {extension_name}: {err}");
+            } else {
+                extensions.push(extension_lock);
+            };
+        }
+
+        rt.extensions = Arc::new(extensions);
+
+        rt
+    }
+
+    #[must_use]
+    pub fn datafusion(&self) -> Arc<RwLock<DataFusion>> {
+        Arc::clone(&self.df)
+    }
+
+    async fn initialize_extension(
```

Comment:
If this is only going to be called once during Runtime::new(), then we don't need a lock - we can just pass a mutable reference to the extension directly - since there aren't any other references.

Reply:
Fixed in https://github.com/spiceai/spiceai/pull/1476/commits/58fccd3e6b01b1a7ed5b5eea399a2e1cdb2e1c4d